pkg/controller.v1/pytorch/env.go

```diff
@@ -0,0 +1,165 @@
+	"strings"
+
+	"github.com/kubeflow/tf-operator/pkg/common/jobcontroller"
+	corev1 "k8s.io/api/core/v1"
+
+	v1 "github.com/kubeflow/pytorch-operator/pkg/apis/pytorch/v1"
+)
+
+const (
+	EnvMasterAddr       = "MASTER_ADDR"
+	EnvMasterPort       = "MASTER_PORT"
+	EnvWorldSize        = "WORLD_SIZE"
+	EnvRank             = "RANK"
+	EnvPythonUnbuffered = "PYTHONUNBUFFERED"
+
+	// EnvPetMasterAddr is the environment variable name for the master address.
+	EnvPetMasterAddr = "PET_MASTER_ADDR"
+	// EnvPetMasterPort is the environment variable name for the master port.
+	EnvPetMasterPort = "PET_MASTER_PORT"
+	// EnvPetNprocPerNode is the environment variable name for the number of processes per node.
+	EnvPetNprocPerNode = "PET_NPROC_PER_NODE"
+	// EnvPetNnodes is the environment variable name for the number of nodes.
+	EnvPetNnodes = "PET_NNODES"
+	// EnvPetNodeRank is the environment variable name for the rank of nodes.
+	EnvPetNodeRank = "PET_NODE_RANK"
+)
+
+// setPodTemplateEnv sets the environment variables for the podTemplateSpec.
+func setPodTemplateEnv(podTemplateSpec *corev1.PodTemplateSpec, job *v1.PyTorchJob, rtype v1.PyTorchReplicaType, index string) error {
+	rank, err := strconv.Atoi(index)
+	if err != nil {
+		return err
+	}
+
+	masterAddr := jobcontroller.GenGeneralName(job.Name, strings.ToLower(string(v1.PyTorchReplicaTypeMaster)), strconv.Itoa(0))
+	if rtype == v1.PyTorchReplicaTypeMaster {
+		if rank != 0 {
+			return errors.New("invalid config: There should be only a single master with index=0")
+		}
+	} else {
+		rank = rank + 1
+	}
+
+	masterPort, err := GetPortFromPyTorchJob(job, v1.PyTorchReplicaTypeMaster)
+	if err != nil {
+		return err
+	}
+
+	for i := range podTemplateSpec.Spec.Containers {
+		masterAddrFound := false
+		masterPortFound := false
+		worldSizeFound := false
+		pythonUnbufferedFound := false
+		petMasterAddrFound := false
+		petMasterPortFound := false
+		petNnodesFound := false
+		petNodeRankFound := false
+
+		for _, env := range podTemplateSpec.Spec.Containers[i].Env {
+			switch env.Name {
+			case EnvMasterAddr:
+				masterAddrFound = true
+			case EnvMasterPort:
+				masterPortFound = true
+			case EnvWorldSize:
+				worldSizeFound = true
+			case EnvPythonUnbuffered:
+				pythonUnbufferedFound = true
+			case EnvPetMasterAddr:
+				petMasterAddrFound = true
+			case EnvPetMasterPort:
+				petMasterPortFound = true
+			case EnvPetNnodes:
+				petNnodesFound = true
+			case EnvPetNodeRank:
+				petNodeRankFound = true
+			}
+		}
+
+		if !masterAddrFound {
+			podTemplateSpec.Spec.Containers[i].Env = append(podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
+				Name:  EnvMasterAddr,
+				Value: masterAddr,
+			})
+		}
+
+		if !masterPortFound {
+			podTemplateSpec.Spec.Containers[i].Env = append(podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
+				Name:  EnvMasterPort,
+				Value: strconv.Itoa(int(masterPort)),
+			})
+		}
+
+		if !worldSizeFound {
+			worldSize := int(getTotalReplicas(job))
```

Comment:
I think worldsize = nproc * replicas , by default nproc is 1.

Reply:
Done. Will read `nprocPerNode` from env in master replica spec and set `worldSize` to `nprocPerNode * totalReplicas`.